cryodrgn/source.py

```diff
@@ -0,0 +1,400 @@
+        if indices is None:
+            indices = np.arange(self.n)
+        elif np.isscalar(indices):
+            indices = np.array([indices])
+        elif isinstance(indices, Iterable):
+            indices = np.array(np.fromiter(indices, int))
+        elif isinstance(indices, slice):
+            start, stop, step = indices.start, indices.stop, indices.step
+            start = start or 0
+            stop = min(stop, self.n) if stop is not None else self.n
+            step = step or 1
+            assert (
+                start >= 0 and stop >= 0 and step >= 0
+            ), "Only positive start/stop/step supported"
+            indices = np.arange(start, stop, step)
+        else:
+            raise TypeError("Unsupported Type for indices")
+
+        assert isinstance(indices, np.ndarray)
+
+        if np.any(indices >= self.n):
+            raise ValueError(f"indices should be < {self.n}")
+
+        # Convert incoming caller indices to indices that this ImageSource will use
+        indices = np.array(self.indices[indices])
+
+        if indices.size == 1:
+            logger.debug(f"ImageSource returning images for index {indices}")
+        else:
+            logger.debug(
+                f"ImageSource returning images for {len(indices)} indices ({indices[0]}..{indices[-1]})"
+            )
+
+        if self.cache:
+            images = self.cache._images(indices)
+        else:
+            images = self._images(indices).astype(self.dtype)
+        return torch.tensor(images)
+
+    def _images(self, indices: np.ndarray) -> np.ndarray:
+        raise NotImplementedError("Subclasses must implement this")
+
+    def chunks(self, chunksize=1000):
+        return ImageSourceIterator(self, chunksize=chunksize)
+
+
+class ArraySource(ImageSource):
+    def __init__(self, data: np.ndarray):
+        if data.ndim == 2:
+            data = data[np.newaxis, ...]
+        nz, ny, nx = data.shape
+        assert ny == nx, "Only square arrays supported"
+        self.data = data
+
+        super().__init__(L=ny, n=nz)
+
+    def _images(self, indices: np.ndarray):
+        return self.data[indices, ...]
+
+
+class MRCFileSource(ImageSource):
+    def __init__(self, filepath: str, *args, **kwargs):
+        from cryodrgn.mrc import MRCHeader
+
+        header = MRCHeader.parse(filepath)
+        self.header = header
+        self.mrcfile_path = filepath
+        self.dtype = header.dtype
+        self.start = 1024 + header.fields["next"]  # start of image data
+        self.nz, self.ny, self.nx = (
+            header.fields["nz"],
+            header.fields["ny"],
+            header.fields["nx"],
+        )
+        assert self.ny == self.nx, "Only square images supported"
+        self.size = self.ny * self.nx
+        self.stride = self.dtype().itemsize * self.size
+
+        super().__init__(L=self.ny, n=self.nz, filenames=filepath, *args, **kwargs)
+
+    def _images(
+        self,
+        indices: np.ndarray,
```

Comment:
You said everything is going to switch to `Tensor` - does that eventually include these fields as well? Or will indices stay as ndarray?

Reply:
In this particular case - `_images`, which subclasses implement, it should remain an `ndarray`, since subclasses that deal with i/o stuff deal with numpy arrays (whether we use standard libraries like `mrcfile` or roll our own implementation like we're doing here). The question is whether to use `Tensor`s as indices in the parent class's `images()` method. I'm inclined to say yes for now, since pytorch indexing seems to work great with numpy indices, and I'm not sure what cascading effects changing those to Tensors will have in `cryoDRGN`. Perhaps we add an issue on this and do a deeper dive on this later? (It _may_ be an additional performance gain..)